File: backend/service/segments.go
package service

import (
	"regexp"
	"strconv"
	"strings"
)

// ReplySegment 完整回复中的一个显式小段
type ReplySegment struct {
	Index   int    `json:"index"`
	SegID   string `json:"seg_id"`
	Content string `json:"content"`
	// Start/End 相对完整原文的 rune 下标（半开区间），便于后续流式打断引用
	Start int `json:"start"`
	End   int `json:"end"`
}

var (
	blankLineRe = regexp.MustCompile(`\n\s*\n+`)
	// 单独成行的显式分段符
	explicitSegRe = regexp.MustCompile(`(?m)^\s*(?:-{3,}|<<<SEG>>>|【段】)\s*$`)
)

const (
	segShortLineMax = 40
	segShortMinLine = 2
)

// SplitReplySegments 把一整段 AI 回复切分为显式小段（确定性规则，不依赖流式打断）。
//
// 规则（v1）：
//  1. 若存在显式分段符（单独一行的 --- / <<<SEG>>> / 【段】），优先按其切开
//  2. 否则按空行（一个及以上）切段
//  3. 若结果仍只有一段且含多行「短句」（≥2 行且每行 ≤40 rune），再按单个 \n 切开
//  4. 丢弃切完后的空白段；空原文返回空切片
//  5. 无显式边界时至少返回 1 段（整段原文）
//
// SegID 使用消息内稳定序号字符串（"0","1",...）；完整引用 id 为 message_id:seg_id。
func SplitReplySegments(content string) []ReplySegment {
	normalized := strings.ReplaceAll(content, "\r\n", "\n")
	normalized = strings.ReplaceAll(normalized, "\r", "\n")
	if strings.TrimSpace(normalized) == "" {
		return nil
	}

	var raws []string
	if explicitSegRe.MatchString(normalized) {
		raws = explicitSegRe.Split(normalized, -1)
	} else {
		raws = blankLineRe.Split(normalized, -1)
	}

	raws = compactSegments(raws)
	if len(raws) <= 1 && shouldSplitShortLines(normalized) {
		raws = compactSegments(strings.Split(normalized, "\n"))
	}
	if len(raws) == 0 {
		return nil
	}

	return buildSegments(normalized, raws)
}

func compactSegments(parts []string) []string {
	var out []string
	for _, p := range parts {
		p = strings.TrimSpace(p)
		if p == "" {
			continue
		}
		out = append(out, p)
	}
	return out
}

func shouldSplitShortLines(normalized string) bool {
	lines := strings.Split(normalized, "\n")
	n := 0
	for _, line := range lines {
		t := strings.TrimSpace(line)
		if t == "" {
			continue
		}
		// 显式分段符行不算内容行
		if explicitSegRe.MatchString(t) {
			continue
		}
		if utf8Len(t) > segShortLineMax {
			return false
		}
		n++
	}
	return n >= segShortMinLine
}

func buildSegments(full string, raws []string) []ReplySegment {
	fullRunes := []rune(full)
	cursor := 0
	out := make([]ReplySegment, 0, len(raws))
	for i, raw := range raws {
		start := indexRunes(fullRunes, []rune(raw), cursor)
		if start < 0 {
			start = cursor
		}
		end := start + utf8Len(raw)
		if end > len(fullRunes) {
			end = len(fullRunes)
		}
		cursor = end
		out = append(out, ReplySegment{
			Index:   i,
			SegID:   strconv.Itoa(i),
			Content: raw,
			Start:   start,
			End:     end,
		})
	}
	return out
}

func indexRunes(haystack, needle []rune, from int) int {
	if len(needle) == 0 {
		return from
	}
	if from < 0 {
		from = 0
	}
	for i := from; i+len(needle) <= len(haystack); i++ {
		match := true
		for j := 0; j < len(needle); j++ {
			if haystack[i+j] != needle[j] {
				match = false
				break
			}
		}
		if match {
			return i
		}
	}
	return -1
}

File: frontend/src/utils/segments.ts
/** AI 回复显式小段（与后端 service.SplitReplySegments 规则一致） */

export interface ReplySegment {
  index: number
  seg_id: string
  content: string
  start?: number
  end?: number
}

const SHORT_LINE_MAX = 40
const SHORT_MIN_LINE = 2
const EXPLICIT_SEG = /^\s*(?:-{3,}|<<<SEG>>>|【段】)\s*$/

function runeLen(s: string): number {
  return Array.from(s).length
}

function compact(parts: string[]): string[] {
  const out: string[] = []
  for (const p of parts) {
    const t = p.trim()
    if (t) out.push(t)
  }
  return out
}

function shouldSplitShortLines(normalized: string): boolean {
  const lines = normalized.split('\n')
  let n = 0
  for (const line of lines) {
    const t = line.trim()
    if (!t) continue
    if (EXPLICIT_SEG.test(t)) continue
    if (runeLen(t) > SHORT_LINE_MAX) return false
    n++
  }
  return n >= SHORT_MIN_LINE
}

function buildSegments(full: string, raws: string[]): ReplySegment[] {
  const fullRunes = Array.from(full)
  let cursor = 0
  const out: ReplySegment[] = []
  raws.forEach((raw, i) => {
    const needle = Array.from(raw)
    let start = -1
    const from = Math.max(0, cursor)
    outer: for (let p = from; p + needle.length <= fullRunes.length; p++) {
      for (let j = 0; j < needle.length; j++) {
        if (fullRunes[p + j] !== needle[j]) continue outer
      }
      start = p
      break
    }
    if (start < 0) start = cursor
    const end = Math.min(fullRunes.length, start + needle.length)
    cursor = end
    out.push({
      index: i,
      seg_id: String(i),
      content: raw,
      start,
      end,
    })
  })
  return out
}

/**
 * 把完整 AI 回复切分为显式小段。
 * 规则与后端一致：显式分段符 > 空行 > 短句换行回退；空段丢弃。
 */
export function splitReplySegments(content: string): ReplySegment[] {
  let normalized = content.replace(/\r\n/g, '\n').replace(/\r/g, '\n')
  if (!normalized.trim()) return []

  let raws: string[]
  if (normalized.split('\n').some((l) => EXPLICIT_SEG.test(l))) {
    raws = normalized.split(/^\s*(?:-{3,}|<<<SEG>>>|【段】)\s*$/m)
  } else {
    raws = normalized.split(/\n\s*\n+/)
  }

  raws = compact(raws)
  if (raws.length <= 1 && shouldSplitShortLines(normalized)) {
    raws = compact(normalized.split('\n'))
  }
  if (raws.length === 0) return []
  return buildSegments(normalized, raws)
}

/** 拼接可引用的段 ID：message_id:seg_id */
export function formatSegRef(messageId: number | string, segId: string): string {
  return `${messageId}:${segId}`
}
